libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/replay/ReplayRouteLocationEngine.java

```diff
@@ -0,0 +1,168 @@
+package com.mapbox.services.android.navigation.v5.location.replay;
+
+import android.annotation.SuppressLint;
+import android.location.Location;
+import android.os.Handler;
+import android.support.annotation.Nullable;
+
+import com.mapbox.android.core.location.LocationEngine;
+import com.mapbox.android.core.location.LocationEngineListener;
+import com.mapbox.api.directions.v5.models.DirectionsRoute;
+import com.mapbox.geojson.LineString;
+import com.mapbox.geojson.Point;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class ReplayRouteLocationEngine extends LocationEngine implements Runnable {
+
+  private static final int HEAD = 0;
+  private ReplayRouteLocationConverter converter;
+  private Handler handler;
+  private List<Location> mockedLocations;
+  private ReplayLocationDispatcher dispatcher;
+  private Location lastLocation = new Location("ReplayRouteLocation");
+  private final ReplayLocationListener replayLocationListener = new ReplayLocationListener() {
+    @Override
+    public void onLocationReplay(Location location) {
+      for (LocationEngineListener listener : locationListeners) {
+        listener.onLocationChanged(location);
+      }
+      lastLocation = location;
+      mockedLocations.remove(HEAD);
+    }
+  };
+
+  public ReplayRouteLocationEngine() {
+    this.handler = new Handler();
+  }
+
+  @SuppressLint("MissingPermission")
+  public void assign(DirectionsRoute route) {
+    handler.removeCallbacks(this);
+    converter = new ReplayRouteLocationConverter(route);
+    mockedLocations = converter.toLocations();
+    dispatcher = obtainDispatcher();
+    dispatcher.run();
+    int mockedPoints = mockedLocations.size();
+    if (mockedPoints <= 5) {
+      handler.postDelayed(this, 1000);
+    } else {
+      handler.postDelayed(this, (mockedPoints - 5) * 1000);
+    }
+  }
+
+  public void assignLastLocation(Point currentPosition) {
+    lastLocation.setLongitude(currentPosition.longitude());
+    lastLocation.setLatitude(currentPosition.latitude());
+  }
+
+  @SuppressLint("MissingPermission")
+  public void moveTo(Point point) {
```

Comment:
There's a lot going on in this `public` method.  Can we extract into some smaller `private` methods to increase readability?